src/components/App.jsx

```diff
@@ -29,10 +37,41 @@ class App extends React.Component {
     if (!props.initialised) {  //NOTE: This should almost always trigger, since App.constructor() triggers exactly once, on the website loading, when all initial values are at their default.
       props.dispatch(checkLoginUser());
     }
+    
+    //SPECIAL: users on ASM tend to stay in a single session for WAY longer
+    //than a standard Zooniverse CFE user, so we're encountering issues where
+    //their login tokens timeout in the middle of classifying a document.
+    //The following is a mechanism that checks if a user is logged in before
+    //performing any action that may require a valid login token, taking safety
+    //actions if necessary.
+    apiClient.beforeEveryRequest = this.checkIfLoggedInUserIsStillLoggedIn.bind(this);
   }
-
-  returnSometrolhing(something) { // eslint-disable-line class-methods-use-this
-    return something;
+  
+  checkIfLoggedInUserIsStillLoggedIn() {
+    const props = this.props;
+    
+    return oauth.checkBearerToken()
+      .then((token) => {
+        //DEBUG/TEST
+        //----------------
+        console.log(`checkBearerToken - initialised: ${!!props.initialised}, user: ${!!props.user}, token: ${!!token}`);
+        const DEBUG_TEST = localStorage.getItem("TEST_TOKEN_ERROR");  //Use localStorage.TEST_TOKEN_ERROR = true to trigger the dialog
+        //----------------
+
+        //If the App thinks you're logged in, but the token says otherwise, deploy emergency measures.
+        if (!!DEBUG_TEST || (props.initialised && props.user && !token)) {          
+          this.props.dispatch(emergencySaveWorkInProgress());
+          this.props.dispatch(toggleDialog(<DialogOfFailure />, false, true));
+          return Promise.reject('User is supposed to be logged in, but token has expired.');  //WARNING: Testing has show that this does NOT abort the Panoptes request.
```

Comment:
[Promise.reject](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Promise/reject) on MDN recommends that the reason be an Error: `return Promise.reject(new Error('Token has expired'));`

I'm not sure why this isn't breaking the promise chain and aborting the API request.

Reply:
Oh SNAAAP it's because the `.catch()` stops the chain from breaking!

```
checkIfLoggedInUserIsStillLoggedIn() {
  return oauth.checkBearerToken()
    .then((token) => {
      return Promise.reject(new Error('pls stop the rest of the API request'));
    })
    .catch((err) => {
      //DERP the Promise.reject from above gets caught here, and the rest of the Promise resolves normally.
    });
}
```

Correct way to kill that API request:
```
checkIfLoggedInUserIsStillLoggedIn() {
  return oauth.checkBearerToken()
    .then((token) => {
      return Promise.reject(new Error('pls stop the rest of the API request'));
    })
    .catch((err) => {
      return Promise.reject(err);  //THIS will absolutely definitely kill the request.
      //Mind you, you might get double error messages logged if the external API request also has console.error().
    });
}
```

Another way to prove that Promise.reject() actually kills the parent API request is to remove that .catch() protection.

Reply:
You could also reverse the order of the try and catch blocks.
```javascript
return oauth.checkBearerToken()
  .catch(error => // catch any errors thrown by the API client)
  .then(token => // rejecting here will block the chain until the next catch block)
```

Reply:
Thinking about it, do you really need to catch errors here? I guess the network might be down, or Panoptes might not be responding to requests, but I don't know if the OAuth module passes those errors back anyway.

@simoneduca this might also be relevant to the work you are doing in https://github.com/zooniverse/shakespeares_world/pull/369